File: torch_tpu/_internal/sync/sync.cc
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "torch_tpu/_internal/sync/sync.h"

#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "ATen/core/TensorBody.h"
#include "absl/base/nullability.h"
#include "absl/container/flat_hash_map.h"
#include "absl/container/flat_hash_set.h"
#include "absl/flags/declare.h"
#include "absl/log/absl_log.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/types/span.h"
#include "mlir/IR/DialectRegistry.h"
#include "mlir/IR/MLIRContext.h"
#include "mlir/Support/DebugStringHelper.h"
#include "torch_tpu/common/error_utils.h"
#include "torch_tpu/common/flags.h"
#include "torch_tpu/eager/device_buffer.h"
#include "torch_tpu/eager/materialize.h"
#include "torch_tpu/eager/structured_log_buffer.h"
#include "torch_tpu/eager/tensor_to_buffer.h"
#include "torch_tpu/eager/traversal.h"
#include "torch_tpu/experimental/eager/materialize_new.h"
#include "xla/hlo/translate/register.h"

ABSL_DECLARE_FLAG(bool, torch_tpu_internal_enable_new_materialization);

namespace torch_tpu {

using BufferRefToVarMap = absl::flat_hash_map<DeviceBufferRef, std::string>;

absl::Status SynchronizeTensors(absl::Span<const at::Tensor> tensors) {
  // Materialize the tensors and get their DeviceBufferRefs.
  // GetMaterialized() will materialize each base tensor and each view.
  TT_ASSIGN_OR_RETURN(
      std::vector<DeviceBufferRef> buffer_refs,
      GetMaterialized(tensors, MaterializationReason::kExplicitSync));

  // Wait for all of the PjRtBuffers to be ready.
  for (const DeviceBufferRef& buffer_ref : buffer_refs) {
    TT_RETURN_IF_ERROR(buffer_ref.Synchronize());
  }

  return absl::OkStatus();
}

absl::Status SynchronizeAll(const WaitOnExecution wait) {
  const std::vector<SharedDeviceBufferList> leaf_nodes =
      SubgraphRegistry::GetInstance().MergeAll()->GetLeafNodes();

  if (leaf_nodes.empty()) {
    return absl::OkStatus();
  }

  TT_RETURN_IF_ERROR(
      Materialize(leaf_nodes, MaterializationReason::kExplicitSync));

  if (GetFlagOnce<bool,
                  &FLAGS_torch_tpu_internal_enable_new_materialization>()) {
    // Ensure that all dispatched graphs have been materialized.
    TT_RETURN_IF_ERROR(BlockOnPendingMaterializations());
  }

  if (wait == WaitOnExecution::kYes) {
    for (const auto& leaf_node : leaf_nodes) {
      TT_RETURN_IF_ERROR(leaf_node->Synchronize());
    }
  }

  return absl::OkStatus();
}

absl::StatusOr<bool> IsMaterialized(const at::Tensor& tensor) {
  // Since view tensors are ephemeral and re-materialized every time, there's
  // no point in checking whether the view is materialized.
  // Instead, we check the base buffer.
  TT_ASSIGN_OR_RETURN(auto base_buffer_ref, GetBaseBuffer(tensor));

  return (base_buffer_ref.state() == DeviceBufferRefState::kMaterialized);
}

absl::StatusOr<bool> IsReady(const at::Tensor& tensor) {
  // Since view tensors are ephemeral and re-materialized every time, there's
  // no point in checking whether the view is ready.
  // Instead, we check the base buffer.
  TT_ASSIGN_OR_RETURN(auto buffer_ref, GetBaseBuffer(tensor));
  if (buffer_ref.state() != DeviceBufferRefState::kMaterialized) {
    // Deferred and placeholder buffers are not ready.
    return false;
  }
  TT_ASSIGN_OR_RETURN(auto* pjrt_buffer, buffer_ref.AwaitBuffer());
  return pjrt_buffer->GetReadyFuture().IsReady();
}

namespace {

absl::StatusOr<absl_nonnull std::unique_ptr<Traversal>> GetTraversal(
    absl::Span<const DeviceBufferRef> buffer_refs) {
  TT_RET_CHECK(!buffer_refs.empty(), error::kInvalidArgument)
      << "tensors must not be empty";
  absl::flat_hash_set<const DeviceBufferList*> nodes_to_traverse_set;
  std::vector<SharedDeviceBufferList> nodes_to_traverse;
  for (const DeviceBufferRef& buffer_ref : buffer_refs) {
    switch (buffer_ref.state()) {
      case DeviceBufferRefState::kMaterialized:
      case DeviceBufferRefState::kDeferred: {
        if (nodes_to_traverse_set.insert(buffer_ref.device_buffer_list().get())
                .second) {
          nodes_to_traverse.push_back(buffer_ref.device_buffer_list());
        }
        continue;
      }
      case DeviceBufferRefState::kPlaceholder:
        return TT_ERROR(error::kInternal)
               << "[GetTraversal] was called on a placeholder "
                  "tensor. This should"
                  " never happen.";
      default:
        return TT_ERROR(error::kInternal)
               << "DeviceBufferRef has unknown state";
    }
  }

  TT_ASSIGN_OR_RETURN(auto traversal, Traversal::Create(nodes_to_traverse));
  ABSL_VLOG(2) << "[GetTraversal] Traversal created";
  return traversal;
}

}  // namespace

absl::StatusOr<std::string> GetComputationGraphviz(
    absl::Span<const DeviceBufferRef> buffer_refs,
    const BufferRefToVarMap& buffer_ref_to_var) {
  TT_ASSIGN_OR_RETURN(auto traversal, GetTraversal(buffer_refs));
  TT_ASSIGN_OR_RETURN(std::string graphviz_string,
                      GetGraphviz(*traversal, buffer_ref_to_var));
  return graphviz_string;
}

absl::StatusOr<std::string> GetComputationMlir(
    absl::Span<const DeviceBufferRef> buffer_refs) {
  TT_ASSIGN_OR_RETURN(auto traversal, GetTraversal(buffer_refs));
  mlir::DialectRegistry registry;
  xla::RegisterMlirToHloDependentDialects(registry);
  mlir::MLIRContext context(registry);
  context.loadAllAvailableDialects();
  TT_ASSIGN_OR_RETURN(auto module, traversal->BuildMlirModule(context));
  return mlir::debugString(module.get());
}

}  // namespace torch_tpu

File: torch_tpu/_internal/sync/sync.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Handles Python type overloading for synchronize() function."""

import inspect
import os

import torch
import torch.distributed.tensor as dt
from torch.utils import _pytree
from torch_tpu._internal.sync import _tpu_torch_sync


def _maybe_unwrap(tensor: torch.Tensor) -> torch.Tensor:
  if isinstance(tensor, dt.DTensor):
    return tensor._local_tensor  # pylint: disable=protected-access
  if hasattr(tensor, "elem"):
    # This is a FakeTensor (or another wrapper). Extract the inner tensor.
    return tensor.elem
  return tensor


def synchronize(
    tensors: torch.Tensor | list[torch.Tensor] | None = None, wait: bool = False
) -> None:
  """Forces a materialization of one or more TPU tensors.

  Args:
    tensors: what to synchronize. If None, synchronizes the entire graph of
      deferred operations. Otherwise, synchronizes only the specified tensor or
      list of tensors.
    wait: Whether to wait for the tensor(s) to be ready. If wait is False, the
      function will compiled a graph to compute the tensors and enqueue it for
      execution, but will not wait for the results. If wait is True, the
      function will also wait for the results to be ready.
  """
  if tensors is None:
    _tpu_torch_sync._synchronize_all(wait)  # pylint: disable=protected-access
  elif isinstance(tensors, list):
    _tpu_torch_sync._synchronize_list(  # pylint: disable=protected-access
        [_maybe_unwrap(t) for t in tensors], wait
    )
  else:
    _tpu_torch_sync._synchronize_tensor(  # pylint: disable=protected-access
        _maybe_unwrap(tensors), wait
    )


def is_materializing(tensor: torch.Tensor) -> bool:
  return _tpu_torch_sync._is_materializing(  # pylint: disable=protected-access
      _maybe_unwrap(tensor)
  )


def is_materialized(tensor: torch.Tensor) -> bool:
  return _tpu_torch_sync._is_materialized(  # pylint: disable=protected-access
      _maybe_unwrap(tensor)
  )


def _dump_computation(
    str_to_dump: str,
    file_name: str,
):
  """Dumps a string to a file."""

  directory = os.path.dirname(file_name)
  if directory:
    os.makedirs(directory, exist_ok=True)
  with open(file_name, "w") as f:
    f.write(str_to_dump)


def dump_computation_graphviz(tensors: list[torch.Tensor], file_name: str):
  """Dumps a graphviz compatible representation of the aten traversal of the given tensors, to a file_name."""
  _dump_computation(computation_graphviz(*tensors), file_name)


def dump_computation_mlir(tensors: list[torch.Tensor], file_name: str):
  """Dumps a graphviz compatible representation of the aten traversal of the given tensors, to a file_name."""
  _dump_computation(computation_mlir(*tensors), file_name)


def computation_graphviz(
    *tensors: list[torch.Tensor] | tuple[torch.Tensor, ...]
) -> str:
  """Returns a graphviz compatible representation of the aten traversal of the given tensors."""
  caller_frame = inspect.stack()[1]  # get the caller's frame
  caller_locals = dict(caller_frame[0].f_locals)
  flat_tensors, _ = _pytree.tree_flatten(tensors)
  return _tpu_torch_sync._get_computation_graphviz(flat_tensors, caller_locals)  # pylint: disable=protected-access


def computation_mlir(
    *tensors: list[torch.Tensor] | tuple[torch.Tensor, ...]
) -> str:
  """Returns a MLIR representation of the aten traversal of the given tensors."""
  flat_tensors, _ = _pytree.tree_flatten(tensors)
  return _tpu_torch_sync._get_computation_mlir(flat_tensors)  # pylint: disable=protected-access
